spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,499 @@
+# * `spam_task_slice:{slice_name}_pred`: a _masked_ set of the original task labels (in this case, labeled `spam_task`) for each slice. Examples that are masked (with `label=-1`) will not contribute to loss or scoring.
+
+# %%
+from snorkel.slicing import add_slice_labels
+
+slice_names = [sf.name for sf in sfs]
+add_slice_labels(dl_train, spam_task, S_train, slice_names)
+add_slice_labels(dl_valid, spam_task, S_valid, slice_names)
+add_slice_labels(dl_test, spam_task, S_test, slice_names)
+
+# %%
+dl_valid.dataset
+
+# %% [markdown]
+# With our updated dataloader, we want to evaluate our model on the defined slice.
+# In the  [`MultitaskClassifier`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.MultitaskClassifier.html), we can call [`score`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.MultitaskClassifier.html#snorkel.classification.MultitaskClassifier.score) with an additional argument, `remap_labels`, to specify that the slice's prediction labels, `spam_task_slice:short_link_pred`, should be mapped to the `spam_task` for evaluation.
+
+# %%
+model.score(
+    dataloaders=[dl_valid, dl_test],
+    remap_labels={"spam_task_slice:short_link_pred": "spam_task"},
+    as_dataframe=True,
+)
+
+# %% [markdown]
+# ### Performance monitoring with [`SliceScorer`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/slicing/snorkel.slicing.SliceScorer.html#snorkel.slicing.SliceScorer)
+
+# %% [markdown]
+# If you're using a model other than [`MultitaskClassifier`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.MultitaskClassifier.html#snorkel-classification-multitaskclassifier), you can still evaluate on slices using the more general `SliceScorer` class.
+#
+# We define a `LogisticRegression` model from sklearn and show how we might visualize these slice-specific scores.
+
+# %%
+from sklearn.linear_model import LogisticRegression
+
+sklearn_model = LogisticRegression(C=0.001, solver="liblinear")
+sklearn_model.fit(X=X_train, y=Y_train)
+sklearn_model.score(X_test, Y_test)
+
+# %% [markdown]
+# Now, we initialize the [`SliceScorer`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/slicing/snorkel.slicing.SliceScorer.html#snorkel.slicing.SliceScorer) using 1) an existing [`Scorer`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/slicing/snorkel.slicing.SliceScorer.html) and 2) desired `slice_names` to see slice-specific performance.
+
+# %%
+from snorkel.utils import preds_to_probs
+from snorkel.slicing import SliceScorer
+
+
+preds_test = sklearn_model.predict(X_test)
+
+scorer = Scorer(metrics=["accuracy", "f1"])
+scorer = SliceScorer(scorer, slice_names)
+scorer.score(
+    S=S_test,
+    golds=Y_test,
+    preds=preds_test,
+    probs=preds_to_probs(preds_test, 2),
+    as_dataframe=True,
+)
+
+# %% [markdown]
+# ## 5. Improve slice performance
```

Comment:
Read in sequence, this kind of sounds like now we're going to improve the performance of the slice we just worked on; rather than adding new slices which is what's happening